beps/bep_0056.rst

```diff
@@ -0,0 +1,145 @@
+This extension adds a capability for clients to negotiate and use
+compression methods for data streams or torrent pieces, effectively
+improving bandwidth for supporting clients.
+
+Rationale
+=========
+This extension would allow clients to download files faster, without
+using file archivers. Since large files are often pre-compressed before
+torrent creation, downloaders needs to keep both the archives
+(for seeding) and uncompressed files (for own usage).
+
+Most users prefer to remove such torrents, thus harming proper file
+distribution. For example: Organizations using Bittorrent for software
+distribution needs to have centralized storage for new customers, no
+matter how many customers have the same software already.
+
+Extension header
+================
+
+This extension uses the extension protocol (specified in `BEP 0010`_)
+to advertise client capability of using chunk compression. It defines
+following items in the extension protocol handshake message:
++-------+-----------------------------------------------------------+
+| name  | description                                               |
++=======+===========================================================+
+| c     | Dictionary of supported compression algorithms which maps |
+|       | its identifiers to its priority (unsigned 8-bit integer), |
+|       | clients can adjust it based on compression speed/ratio,   |
+|       | hardware support, performance, and power mode et cetera.  |
+|       | Priority set to zero means that the compression algorithm |
+|       | is not supported or disabled by user, the client must     |
+|       | ignore unknown algorithms.                                |
++-------+-----------------------------------------------------------+
+
+
+
+The compression algorithm is selected by taking the dictionary item with
+highest priority from intersection of items supported by both peers,
+if there isn't any suitable compression algorithm - compression will be disabled.
+
+Example of extension handshake message:
+
+::
+
+  {
+    'c': {
+      'p_zstd': 255,
+      's_zstd': 153,
+      'p_lz4': 106,
+      'p_density': 70,
+      's_lz4': 41,
+      's_density': 37
+    }
+  }
+
+
+Compression methods
+===================
+Extension provides two approaches (methods) to compression, which have
+their own trade-offs, so choice between these should be made by clients
+on per-torrent basis, using its metadata (properties like piece size).
+
+With **by-piece compression** method, client must compress each piece
+individually, which lowers overall compression ratio but result can
+be stored in cache and reused, probably providing more efficiency.
+If the client is caching compressed pieces in memory, then it can be
+decompressed when saving to disk or sending to peer, which not supports
+compression. To reduce piece re-compression, client should raise
+current algorithm's priority during handshake. This method has low
+efficiency with pieces smaller than 4 MB.
```

Comment:
There are a lot of details omitted here. This needs to fit into the way blocks are requested and sent according to the protocol, see http://bittorrent.org/beps/bep_0003.html

Crucially, when you say the whole piece is compressed, do you mean that I have to request *all* blocks for that piece from the same peer, in order to decompress any part of it?

The offset and size that's specified in the request message, is the referring to the uncompressed piece (as it does in the current protocol) or does it refer to the compressed piece? The requestor would need to know the compressed size of each piece in that case, which there doesn't seem to be a mechanism to learn.

It seems far more practical to introduce a new `PIECE` message which indicates which compression algorithm it's using, leaving everything else the same. But that would require compressing each block individually, and maybe even smaller and unaligned parts of pieces. You don't have to request blocks at 16 kiB alignments.

Reply:
Thank you, I should have introduced `CPIECE` message in the first place.